<a href="https://colab.research.google.com/github/hemant99108/deep-learning/blob/main/ANN/batchNormalizationLayer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **observe the results before and after the batch normalization**

# **before applying batch normalization**

In [2]:
#importing the necessary libraries
import numpy as np
import pandas as pd
import os
import tensorflow as tf
import matplotlib.pyplot as plt
import seaborn as sns
import time
plt.style.use('fivethirtyeight')
%load_ext tensorboard

In [3]:
#loading the data of fashion mnist
(X_train_full,y_train_full),(X_test,y_test)=tf.keras.datasets.fashion_mnist.load_data()
X_train_full=X_train_full/255.0
X_test=X_test/255.0

X_valid,X_train=X_train_full[:5000],X_train_full[5000:]
y_valid,y_train=y_train_full[:5000],y_train_full[5000:]



4422102/4422102 [==============================] - 0s 0us/step


In [4]:
#creating layers of the model

tf.random.set_seed(42) #for getting similar output optional
np.random.seed(42) #for getting similar outuput optional

LAYERS=[
    tf.keras.layers.Flatten(input_shape=[28,28]),
    tf.keras.layers.Dense(300,kernel_initializer='he_normal'),
    tf.keras.layers.LeakyReLU(),
    tf.keras.layers.Dense(100,kernel_initializer='he_normal'),
    tf.keras.layers.LeakyReLU(),
    tf.keras.layers.Dense(10,activation='softmax')
]

model=tf.keras.models.Sequential(LAYERS)


In [5]:
#compiling the model

model.compile(loss='sparse_categorical_crossentropy',
              optimizer=tf.keras.optimizers.SGD(lr=1e-3),
              metrics=['accuracy'])

In [6]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense (Dense)               (None, 300)               235500    
                                                                 
 leaky_re_lu (LeakyReLU)     (None, 300)               0         
                                                                 
 dense_1 (Dense)             (None, 100)               30100     
                                                                 
 leaky_re_lu_1 (LeakyReLU)   (None, 100)               0         
                                                                 
 dense_2 (Dense)             (None, 10)                1010      
                                                                 
Total params: 266610 (1.02 MB)
Trainable params: 266610 

In [7]:
#now training and calculating the trainning time

#starting time

start=time.time()

history=model.fit(X_train,y_train,epochs=10,
                  validation_data=(X_valid,y_valid),verbose=2)

end=time.time()

#totall time taken
print(f'runtime of the program is {end-start}')

Epoch 1/10
1719/1719 - 8s - loss: 0.7101 - accuracy: 0.7611 - val_loss: 0.5224 - val_accuracy: 0.8264 - 8s/epoch - 4ms/step
Epoch 2/10
1719/1719 - 7s - loss: 0.4930 - accuracy: 0.8284 - val_loss: 0.4415 - val_accuracy: 0.8520 - 7s/epoch - 4ms/step
Epoch 3/10
1719/1719 - 7s - loss: 0.4497 - accuracy: 0.8432 - val_loss: 0.5595 - val_accuracy: 0.7948 - 7s/epoch - 4ms/step
Epoch 4/10
1719/1719 - 7s - loss: 0.4262 - accuracy: 0.8514 - val_loss: 0.4066 - val_accuracy: 0.8626 - 7s/epoch - 4ms/step
Epoch 5/10
1719/1719 - 6s - loss: 0.4102 - accuracy: 0.8567 - val_loss: 0.3922 - val_accuracy: 0.8624 - 6s/epoch - 4ms/step
Epoch 6/10
1719/1719 - 8s - loss: 0.3936 - accuracy: 0.8624 - val_loss: 0.3855 - val_accuracy: 0.8700 - 8s/epoch - 4ms/step
Epoch 7/10
1719/1719 - 7s - loss: 0.3830 - accuracy: 0.8661 - val_loss: 0.3841 - val_accuracy: 0.8710 - 7s/epoch - 4ms/step
Epoch 8/10
1719/1719 - 7s - loss: 0.3730 - accuracy: 0.8670 - val_loss: 0.3989 - val_accuracy: 0.8612 - 7s/epoch - 4ms/step
Epoch 9/

# **runtime of the program is 82.44831538200378**

# **accuracy: 0.8945**

In [8]:
#delete the previous model
del model


In [9]:
#defining the new model with batch normalization
LAYERS_BN=[
    tf.keras.layers.Flatten(input_shape=[28,28]),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(300,activation='relu'),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(100,activation='relu'),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(10,activation='softmax')
]


model=tf.keras.models.Sequential(LAYERS_BN)

In [10]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_1 (Flatten)         (None, 784)               0         
                                                                 
 batch_normalization (Batch  (None, 784)               3136      
 Normalization)                                                  
                                                                 
 dense_3 (Dense)             (None, 300)               235500    
                                                                 
 batch_normalization_1 (Bat  (None, 300)               1200      
 chNormalization)                                                
                                                                 
 dense_4 (Dense)             (None, 100)               30100     
                                                                 
 batch_normalization_2 (Bat  (None, 100)              

In [11]:
bn1=model.layers[1]

In [12]:
for variable in bn1.variables:
  print(variable.name,variable.trainable)

batch_normalization/gamma:0 True
batch_normalization/beta:0 True
batch_normalization/moving_mean:0 False
batch_normalization/moving_variance:0 False


In [13]:
model.compile(loss='sparse_categorical_crossentropy',
              optimizer=tf.keras.optimizers.SGD(lr=1e-3),
               metrics=['accuracy'])

In [14]:
#now train and calculate the training time

#starting time
start=time.time()

history=model.fit(X_train,y_train,epochs=10,
                  validation_data=(X_valid,y_valid),verbose=2)

#ending time
end=time.time()

#total time taken
print(f'total runtime taken is {end-start}')

Epoch 1/10
1719/1719 - 12s - loss: 0.5310 - accuracy: 0.8152 - val_loss: 0.3887 - val_accuracy: 0.8670 - 12s/epoch - 7ms/step
Epoch 2/10
1719/1719 - 10s - loss: 0.3923 - accuracy: 0.8591 - val_loss: 0.3504 - val_accuracy: 0.8744 - 10s/epoch - 6ms/step
Epoch 3/10
1719/1719 - 10s - loss: 0.3537 - accuracy: 0.8727 - val_loss: 0.3412 - val_accuracy: 0.8796 - 10s/epoch - 6ms/step
Epoch 4/10
1719/1719 - 11s - loss: 0.3247 - accuracy: 0.8818 - val_loss: 0.3246 - val_accuracy: 0.8844 - 11s/epoch - 6ms/step
Epoch 5/10
1719/1719 - 13s - loss: 0.3049 - accuracy: 0.8889 - val_loss: 0.3080 - val_accuracy: 0.8926 - 13s/epoch - 8ms/step
Epoch 6/10
1719/1719 - 12s - loss: 0.2889 - accuracy: 0.8949 - val_loss: 0.3130 - val_accuracy: 0.8890 - 12s/epoch - 7ms/step
Epoch 7/10
1719/1719 - 11s - loss: 0.2744 - accuracy: 0.8992 - val_loss: 0.3105 - val_accuracy: 0.8898 - 11s/epoch - 7ms/step
Epoch 8/10
1719/1719 - 11s - loss: 0.2625 - accuracy: 0.9044 - val_loss: 0.3146 - val_accuracy: 0.8870 - 11s/epoch - 7

# ***runtime of the program is 143.8s1***

# **accuracy is 0.89**
# note: as we can see that although it has been taken a long time due to some extra parameter added but it has improved the acccuracy score after applying BN in big problems it performs more